### PR TITLE
Design: Refine code generation APIs

### PR DESCRIPTION
**Pull request by bricelam**
This removes the concept of the programming language from the provider APIs.

### CODE REVIEW

File: test/EFCore.Relational.Tests/TestUtilities/TestProviderScaffoldingCodeGenerator.cs [hunk 1]
```diff
@@ -5,9 +5,9 @@
 
 namespace Microsoft.EntityFrameworkCore.TestUtilities
 {
-    public class TestProviderScaffoldingCodeGenerator : IScaffoldingProviderCodeGenerator
+    public class TestProviderScaffoldingCodeGenerator : IScaffoldingProviderCodeGenerator2
```
Comment (bricelam) [inline test/EFCore.Relational.Tests/TestUtilities/TestProviderScaffoldingCodeGenerator.cs]: Will rename 😏

File: src/EFCore.Relational/Scaffolding/IScaffoldingProviderCodeGenerator2.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.EntityFrameworkCore.Scaffolding
+{
+    /// <summary>
+    ///     A service typically implemented by database providers to generate code fragments
+    ///     for reverse engineering.
+    /// </summary>
+    public interface IScaffoldingProviderCodeGenerator2
```
Comment (bricelam) [inline src/EFCore.Relational/Scaffolding/IScaffoldingProviderCodeGenerator2.cs]: Will rename 😏

File: src/EFCore.Relational/Design/MethodCallCodeFragment.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        public MethodCallCodeFragment([NotNull] string method, [NotNull] params object[] arguments)
+        {
+            Check.NotEmpty(method, nameof(method));
+            Check.NotNull(arguments, nameof(arguments));
+
+            Method = method;
+
+            foreach (var argument in arguments)
+            {
+                Arguments.Add(argument);
```
Comment (smitpatel) [inline src/EFCore.Relational/Design/MethodCallCodeFragment.cs]: AddRange or ctor args to create the list.

Comment (bricelam) [reply src/EFCore.Relational/Design/MethodCallCodeFragment.cs]: IList doesn't have AddRange, and I couldn't see an already existing extension method.

**Comment by bricelam** [comment]
Looks like I need to update Oracle more too...

**Comment by bricelam** [comment]
>  Provider can no longer generate something custom based on connection string.

Yeah, I debated whether to use MethodCallCodeFragment here too...

**Comment by smitpatel** [comment]
I was thinking more like, for SqlServer provider, provider can use connection string to find out server version and call those additional things like
`UseSqlServer("connectionString").UseRowNumberPaging()`
May be a list of MethodCallFragment can represent it.

**Comment by bricelam** [comment]
We'll YAGNI it for now. Now we can actually evolve this interface without breaking providers. 😉

**Comment by bricelam** [comment]
Filed #10487

File: src/EFCore.Design/Scaffolding/Internal/ScaffoldingServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -26,6 +26,7 @@ public static class ScaffoldingServiceCollectionExtensions
         public static IServiceCollection AddScaffolding([NotNull] this IServiceCollection serviceCollection, [NotNull] IOperationReporter reporter)
             => serviceCollection
                 .AddSingleton<AnnotationCodeGeneratorDependencies>()
+                .AddSingleton<CoreTypeMapperDependencies>()
```
Comment (bricelam) [inline src/EFCore.Design/Scaffolding/Internal/ScaffoldingServiceCollectionExtensions.cs]: We really need to add more RevEng tests. (#9111) I'll try to get to it soon.
cc @ajcvickers

**Pull request merged**